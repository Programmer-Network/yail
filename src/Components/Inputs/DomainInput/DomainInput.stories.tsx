import { action } from "@storybook/addon-actions";
import { useEffect, useState } from "react";

import DomainInput from ".";

export default {
  title: "Input/DomainInput",
  component: DomainInput,
  parameters: {
    layout: "centered"
  },
  argTypes: {
    onChange: { action: "changed" }
  }
};

export const Default = () => {
  const [input, setInput] = useState<Record<string, string>>({ url: "" });

  useEffect(() => {
    action("onChange")(input["url"]);
  }, [input["url"]]);

  return (
    <DomainInput
      domain='programmer.network'
      name='url'
      value={input["url"] ?? ""}
      onChange={setInput}
    />
  );
};
