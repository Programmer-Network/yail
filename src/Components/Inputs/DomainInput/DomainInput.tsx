import classNames from "classnames";
import { ChangeEvent, FC, useState } from "react";

import { InputError } from "../Common/InputError";
import { InputHeader } from "../Common/InputHeader";
import { IDomainInputProps } from "./types";

const DomainInput: FC<IDomainInputProps> = (
  props = {
    domain: "",
    name: "",
    value: "",
    hint: "",
    className: "",
    domainClassName: "",
    inputWrapperClassName: "",
    type: "string",
    max: 0,
    error: "",
    disabled: false,
    onChange: () => null
  }
) => {
  const { domain, name, value, disabled } = props;
  const [isFocused, setIsFocused] = useState(false);

  const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
    props.onChange({ [e.target.name]: e.target.value });
  };

  return (
    <div className={classNames("group text-text", props.inputWrapperClassName)}>
      <InputHeader {...props} />
      <div className='flex w-full'>
        <span
          data-testid='domain'
          className={classNames(
            "border-border selection:text-text min-w-max rounded-tl-md rounded-bl-md border-2 px-4 py-2 leading-5 tracking-tighter",
            props.domainClassName,
            {
              "group-hover:border-border": !props.disabled,
              "border-border! ring-3-transparent!": isFocused
            }
          )}
        >
          {domain}
        </span>
        <input
          data-testid='input'
          name={name}
          value={value}
          disabled={disabled}
          onFocus={() => setIsFocused(true)}
          onBlur={() => setIsFocused(false)}
          className={classNames(
            "border-border focus:border-border focus:ring-3-transparent w-full cursor-pointer rounded-tr-md rounded-br-md border-t-2 border-r-2 border-b-2 bg-transparent p-2 transition-colors focus:outline-hidden",
            props.className,
            {
              "group-hover:border-border": !props.disabled,
              "hover:cursor-not-allowed": props.disabled
            }
          )}
          onChange={handleChange}
        />
      </div>
      {props.error && <InputError error={props.error} />}
    </div>
  );
};

export { DomainInput };
