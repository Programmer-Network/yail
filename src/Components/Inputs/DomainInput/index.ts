export { default } from "./DomainInput";
export * from "./types";
