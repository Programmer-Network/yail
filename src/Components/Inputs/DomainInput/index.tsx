import classNames from "classnames";
import { ChangeEvent, FC, useState } from "react";

import InputError from "Components/Inputs/Common/InputError";
import InputHeader from "Components/Inputs/Common/InputHeader";

import { IDomainInputProps } from "./types";

const DomainInput: FC<IDomainInputProps> = (
  props = {
    domain: "",
    name: "",
    value: "",
    hint: "",
    className: "",
    domainClassName: "",
    inputWrapperClassName: "",
    type: "string",
    max: 0,
    error: "",
    disabled: false,
    onChange: () => null
  }
) => {
  const { domain, name, value, disabled } = props;
  const [isFocused, setIsFocused] = useState(false);

  const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
    props.onChange({ [e.target.name]: e.target.value });
  };

  return (
    <div
      className={classNames(
        "group yl:text-secondary",
        props.inputWrapperClassName
      )}
    >
      <InputHeader {...props} />
      <div className='yl:flex yl:w-full'>
        <span
          data-testid='domain'
          className={classNames(
            "yl:min-w-max yl:border-2 yl:border-border yl:px-4 yl:py-2 yl:leading-5 yl:tracking-tighter yl:selection:text-secondary yl:rounded-tl-md yl:rounded-bl-md",
            props.domainClassName,
            {
              "yl:group-hover:border-border": !props.disabled,
              "yl:border-border! !ring-3-transparent": isFocused
            }
          )}
        >
          {domain}
        </span>
        <input
          data-testid='input'
          name={name}
          value={value}
          disabled={disabled}
          onFocus={() => setIsFocused(true)}
          onBlur={() => setIsFocused(false)}
          className={classNames(
            "yl:w-full yl:cursor-pointer yl:border-t-2 yl:border-b-2 yl:border-r-2 yl:border-border yl:bg-transparent yl:p-2 transition-colors yl:focus:border-border yl:focus:outline-hidden yl:focus:ring-3-transparent yl:rounded-tr-md yl:rounded-br-md",
            props.className,
            {
              "yl:group-hover:border-border": !props.disabled,
              "yl:hover:cursor-not-allowed": props.disabled
            }
          )}
          onChange={handleChange}
        />
      </div>
      {props.error && <InputError error={props.error} />}
    </div>
  );
};

export default DomainInput;
