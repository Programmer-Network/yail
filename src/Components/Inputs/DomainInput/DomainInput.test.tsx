import { fireEvent, render, screen } from "@testing-library/react";
import { describe, expect, it, vi } from "vitest";

import DomainInput from "./";

describe("DomainInput component", () => {
  const mockOnChange = vi.fn();

  it("matches snapshot", () => {
    const props = {
      domain: "example.com",
      name: "domainInput",
      value: "test",
      onChange: mockOnChange
    };

    const { asFragment } = render(<DomainInput {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders correctly with initial props", () => {
    const props = {
      domain: "example.com",
      name: "domainInput",
      value: "test",
      onChange: mockOnChange
    };

    render(<DomainInput {...props} />);

    expect(screen.getByDisplayValue(props.value)).toBeInTheDocument();
    expect(screen.getByText(props.domain)).toBeInTheDocument();
  });

  it("calls onChange handler when input value changes", () => {
    const props = {
      domain: "example.com",
      name: "domainInput",
      value: "test",
      onChange: mockOnChange
    };

    render(<DomainInput {...props} />);
    const inputElement = screen.getByDisplayValue(props.value);

    fireEvent.change(inputElement, { target: { value: "newTest" } });
    expect(mockOnChange).toHaveBeenCalledWith({ [props.name]: "newTest" });
  });

  it("applies focus and blur styles", () => {
    const props = {
      domain: "example.com",
      name: "domainInput",
      value: "test",
      onChange: mockOnChange
    };

    render(<DomainInput {...props} />);
    const domain = screen.getByTestId("domain");
    expect(domain).not.toHaveClass("border-border! ring-3-transparent!");

    const input = screen.getByTestId("input");
    fireEvent.focus(input);

    expect(domain).toHaveClass("border-border! ring-3-transparent!");

    fireEvent.blur(input);
    expect(domain).not.toHaveClass("border-border! ring-3-transparent!");
  });

  it("displays error when error prop is provided", () => {
    const props = {
      domain: "example.com",
      name: "domainInput",
      value: "test",
      onChange: mockOnChange,
      error: "Error message"
    };

    render(<DomainInput {...props} />);
    expect(screen.getByText(props.error)).toBeInTheDocument();
  });

  it("applies disabled styles when disabled", () => {
    const props = {
      domain: "example.com",
      name: "domainInput",
      value: "test",
      onChange: mockOnChange,
      disabled: true
    };

    render(<DomainInput {...props} />);
    const inputElement = screen.getByDisplayValue(props.value);
    expect(inputElement).toBeDisabled();
  });
});
