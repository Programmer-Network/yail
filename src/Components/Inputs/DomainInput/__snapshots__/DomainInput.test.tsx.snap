// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DomainInput component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="group yl-text-secondary"
  >
    <div
      class="yl-flex yl-w-full"
    >
      <span
        class="yl-min-w-max yl-border-2 yl-border-border yl-px-4 yl-py-2 yl-leading-5 yl-tracking-tighter selection:yl-text-secondary yl-rounded-tl-md yl-rounded-bl-md group-hover:yl-border-border"
        data-testid="domain"
      >
        example.com
      </span>
      <input
        class="yl-w-full yl-cursor-pointer yl-border-t-2 yl-border-b-2 yl-border-r-2 yl-border-border yl-bg-transparent yl-p-2 transition-colors focus:yl-border-border focus:yl-outline-none focus:yl-ring-transparent yl-rounded-tr-md yl-rounded-br-md group-hover:yl-border-border"
        data-testid="input"
        name="domainInput"
        value="test"
      />
    </div>
  </div>
</DocumentFragment>
`;
